### PR TITLE
CSHARP-3815: Make SemaphoreSlimSignalable.Signal end SemaphoreSlimSignalable.WaitSignaledAsync immediately

### PR DESCRIPTION
**Pull request by BorisDog**
[EG](https://evergreen.mongodb.com/version/61267b51d6d80a4812670e78)

### CODE REVIEW

**Comment by JamesKovacs** [comment]
Discussed avoiding `Task.Yield()` due to possible problems with yielding on UI or ASP thread. The new technique instead of `Task.Yield()` seems kosher to me.

LGTM.

File: tests/MongoDB.Driver.Core.Tests/Core/Misc/SemaphoreSlimSignalableTests.cs [hunk 1]
```diff
@@ -120,6 +161,65 @@ public void SemaphoreSlimSignalable_constructor_should_check_arguments([Values(-
             }
         }
 
+        [Fact]
+        public async Task WaitSignaledAsync_should_not_continue_on_signal_thread()
```
Comment (DmitryLukyanov) [inline tests/MongoDB.Driver.Core.Tests/Core/Misc/SemaphoreSlimSignalableTests.cs:165]: this test is flaky, you may see it if run this test few times in row:

>WaitSignaledAsync_should_not_continue_on_signal_thread([Range(1,10)] int _)

File: src/MongoDB.Driver.Core/Core/Misc/TaskExtensions.cs [hunk 2]
```diff
@@ -13,17 +13,48 @@
...
 
 namespace MongoDB.Driver.Core.Misc
 {
     internal static class TaskExtensions
     {
+        internal struct YieldNoContextAwaitable
+        {
+            public YieldNoContextAwaiter GetAwaiter() { return new YieldNoContextAwaiter(); }
+
+            public struct YieldNoContextAwaiter : ICriticalNotifyCompletion
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver.Core/Core/Misc/TaskExtensions.cs:28]: note: this looks like something similar to what you did for `alwaysYield=false`: https://github.com/microsoft/vs-threading/blob/b6e731c42285fabcec0d900fb32ac25dc1f53fb5/src/Microsoft.VisualStudio.Threading/AwaitExtensions.cs#L458

**Review by DmitryLukyanov — APPROVED** [review]
LGTM

File: src/MongoDB.Driver.Core/Core/Misc/TaskExtensions.cs (hunk 2 above)
Comment (BorisDog) [reply src/MongoDB.Driver.Core/Core/Misc/TaskExtensions.cs:28]: Great find! I hope `AwaitExtensions` will be promoted to `System.Threading` sometime.

**Comment by BorisDog** [comment]
[Merged](https://github.com/mongodb/mongo-csharp-driver/commit/b961b81cb7dc1ffe7262c55a227afad0aab5a994)

**Pull request closed without merging**